src/cswinrt/strings/WinRT.cs

```diff
@@ -1063,110 +1098,105 @@ private EventRegistrationToken AddEventHandlerNoLock(T handler)
+                // When on .NET 6+, just iterate on TryAdd, which allows skipping the extra
+                // lookup on the last iteration (as the handler is added rigth away instead).
+                //
+                // We're doing this do-while loop here and incrementing 'm_low32Bits' on every failed insertion to work
+                // around one possible (theoretical) performance problem. Suppose the candidate token was somehow already
+                // used (not entirely clear when that would happen in practice). Incrementing only the local value from the
+                // loop would mean we could "race past" the value in 'm_low32Bits', meaning that all subsequent registrations
+                // would then also go through unnecessary extra lookups as the value of those lower 32 bits "catches up" to
+                // the one that ended up being used here. So we can avoid that by simply incrementing both of them every time.
+                tokenLow32Bits = m_low32Bits++;
             }
-            else
+            while (!m_tokens.TryAdd(tokenLow32Bits, handler));
+#else
+            do
             {
-                handlerHashCode = (uint)handler.GetHashCode();
+                tokenLow32Bits = m_low32Bits++;
             }
-
-            ulong tokenValue = ((ulong)(uint)typeof(T).GetHashCode() << 32) | handlerHashCode;
-            return new EventRegistrationToken { Value = (long)tokenValue };
+            while (m_tokens.ContainsKey(tokenLow32Bits));
+            m_tokens[tokenLow32Bits] = handler;
+#endif
+            // The real event registration token is composed this way:
+            //   - The upper 32 bits are the hashcode of the T type argument.
+            //   - The lower 32 bits are the valid token computed above.
+            return new EventRegistrationToken { Value = (long)(((ulong)(uint)TypeOfTHashCode << 32) | (uint)tokenLow32Bits) };
         }
 
         // Remove the event handler from the table and
         // Get the delegate associated with an event registration token if it exists
         // If the event registration token is not registered, returns false
         public bool RemoveEventHandler(EventRegistrationToken token, out T handler)
         {
+            // If the token doesn't have the upper 32 bits set to the hashcode of the delegate
+            // type in use, we know that the token cannot possibly have a registered handler.
+            if ((int)((ulong)token.Value >> 32) != TypeOfTHashCode)
+            {
+                handler = null;
+
+                return false;
+            }
+
             lock (m_tokens)
             {
 #if NET6_0_OR_GREATER
                 // On .NET 6 and above, we can use a single lookup to both check whether the token
                 // exists in the table, remove it, and also retrieve the removed handler to return.
-                if (m_tokens.Remove(token, out handler))
+                if (m_tokens.Remove((int)token.Value, out object obj))
```

Comment:
Consider a comment here to indicate the cast to `int` means it uses the lower 32 bits which is what is tracked in the dictionary.